File: Device/Noban/ble_provisioning.py
"""Phase 7 — BLE Wi-Fi + identity provisioning sidecar.

Spawned by main.cpp's ButtonGesture::ENTER_SETUP path when the user holds
the button for 3-10 seconds. Replaces the legacy AP-mode `wifi_setup.sh`
(which forced the user's phone to disconnect from their LAN to hit the
device at 192.168.4.1:8080 — terrible UX).

Flow:
    1. Advertise a GATT service so the companion app (flutter_blue_plus)
       can find us under "ElderlyCare-Setup".
    2. App connects + pairs (Just Works — anyone within BLE range during
       the setup window is implicitly trusted; window is short so the
       attack surface is narrow).
    3. App calls `wifi_scan` (READ) → we run `nmcli dev wifi list`, return
       JSON list of {ssid, signal, security}.
    4. App writes `wifi_join` with chosen network → we run
       `nmcli dev wifi connect`, notify `status` with the result.
    5. App writes `commit_setup` with the user-picked device_id +
       32-char password → we write /var/lib/elderly-care/id.txt, notify
       `done`, exit cleanly.

Topology choice — why a separate Python sidecar instead of doing BLE in
C++ directly: BlueZ's GATT API is DBus-based, and the Python `bless`
library wraps it in a tiny async loop. The C++ equivalent (bluetooth-dev
+ glib + dbus) is 10× the code and a build-time dependency we don't need
99% of the time (device only runs BLE during the ~3-minute setup window).
Sidecar is spawned + killed by main.cpp; no process running otherwise.

Window timing:
    - Sidecar exits after 5 minutes if no client connects (Window A).
    - Once a client is connected, no timeout (Window B) — user can take
      as long as they want to type their Wi-Fi password.
    - Sidecar exits voluntarily after commit_setup completes.
    - main.cpp can also SIGTERM the sidecar process at any time.

Reads/writes id.txt at /var/lib/elderly-care/id.txt (production) or
$ELDERLY_ID_FILE (override for testing). Same path that main.cpp's
loadDeviceIdentity() reads from.
"""
from __future__ import annotations

import asyncio
import json
import logging
import os
import signal
import subprocess
import sys
from pathlib import Path
from typing import Any, Optional

try:
    from bless import (
        BlessServer,
        BlessGATTCharacteristic,
        GATTCharacteristicProperties,
        GATTAttributePermissions,
    )
except ImportError:
    print("[ble] FATAL: `bless` not installed. Run: pip install bless", file=sys.stderr)
    sys.exit(2)


log = logging.getLogger("ble_provisioning")
logging.basicConfig(
    level=logging.INFO,
    format="[ble] %(asctime)s %(levelname)s %(message)s",
)

# ── UUIDs ──────────────────────────────────────────────────────────────────
# Custom 128-bit base; characteristics use sequential suffixes. Companion
# app's BleProvisioningService must use these exact strings.
SVC_UUID         = "11111111-2222-3333-4444-555555555555"
CHAR_WIFI_SCAN   = "11111111-2222-3333-4444-555555555001"  # READ
CHAR_WIFI_JOIN   = "11111111-2222-3333-4444-555555555002"  # WRITE
CHAR_STATUS      = "11111111-2222-3333-4444-555555555003"  # NOTIFY
CHAR_COMMIT      = "11111111-2222-3333-4444-555555555004"  # WRITE
CHAR_DONE        = "11111111-2222-3333-4444-555555555005"  # NOTIFY

def _factory_serial_suffix() -> str:
    """A short factory-unique tag so multiple un-provisioned devices in range
    are distinguishable in the app's picker. Uses the last 4 hex of the Pi's
    CPU serial (stable per board, available before provisioning). Falls back
    to the BT MAC suffix, then a random tag."""
    try:
        for line in Path("/proc/cpuinfo").read_text().splitlines():
            if line.startswith("Serial"):
                s = line.split(":")[-1].strip()
                if len(s) >= 4:
                    return s[-4:].upper()
    except Exception:
        pass
    try:
        import uuid
        return f"{uuid.getnode() & 0xFFFF:04X}"
    except Exception:
        return "0000"

# Advertised as "Noban-Setup-<serial>" so the app can list + pick a specific
# device when several are advertising at once. The service UUID (below) is
# what the app filters on; the suffix is just the human label. Must match the
# app's devicePrefix in ble_provisioning.dart.
DEVICE_NAME      = "Noban-Setup-" + _factory_serial_suffix()

# Where id.txt lives. Must match loadDeviceIdentity() in main.cpp.
ID_FILE = Path(os.environ.get(
    "ELDERLY_ID_FILE", "/var/lib/elderly-care/id.txt"))

# How long to wait for an app to connect before giving up.
WINDOW_A_SECONDS = 5 * 60

# Module-level state. Bless's read/write callbacks are sync (called by
# the dbus loop) but our nmcli ops are blocking, so we shell out and let
# them run in an executor.
_server: Optional[BlessServer] = None
_status_value = b"idle"
_done_value   = b""
_client_connected = False
_setup_completed  = False
_loop: Optional[asyncio.AbstractEventLoop] = None


# ── Helpers ────────────────────────────────────────────────────────────────

def _run_nmcli(*args: str, timeout: int = 30) -> tuple[int, str, str]:
    """Run nmcli, return (rc, stdout, stderr). Doesn't raise on non-zero.

    Runs via `sudo -n` because creating/activating system connection profiles
    (`connection add/up/delete`) needs root — the sidecar runs as the
    unprivileged pipeline user, which otherwise hits NetworkManager's
    'Insufficient privileges'. The pi user has passwordless sudo on Pi OS;
    setup.sh also installs a NOPASSWD rule scoped to /usr/bin/nmcli. If sudo
    is unavailable we fall back to a bare nmcli (works for read-only ops)."""
    try:
        r = subprocess.run(
            ["sudo", "-n", "nmcli", *args],
            capture_output=True, text=True, timeout=timeout,
        )
        # If sudo itself failed (no privilege), retry without it so read-only
        # commands (dev wifi list) still work.
        if r.returncode != 0 and "sudo:" in (r.stderr or ""):
            r = subprocess.run(["nmcli", *args],
                               capture_output=True, text=True, timeout=timeout)
        return r.returncode, r.stdout, r.stderr
    except subprocess.TimeoutExpired:
        return 124, "", "timeout"
    except FileNotFoundError:
        return 127, "", "nmcli not installed"


def _scan_wifi() -> list[dict]:
    """Return list of {ssid, signal, security} for nearby networks. Filters
    duplicates (NM lists each BSSID separately; we keep the strongest)."""
    # --rescan no: return NetworkManager's CACHED scan (instant). A fresh
    # "--rescan yes" blocks for up to ~20 s, which stalls this synchronous BLE
    # read handler past the phone's read timeout → app sees an empty list. NM
    # rescans on its own every ~2 min, and the radio scans on connect, so the
    # cache is reliably populated. Worst case the user taps Scan again.
    rc, out, _err = _run_nmcli(
        "-t", "-f", "SSID,SIGNAL,SECURITY", "dev", "wifi", "list",
        "--rescan", "no", timeout=10,
    )
    if rc != 0:
        return []
    best: dict[str, dict] = {}
    for line in out.strip().splitlines():
        # nmcli -t format: SSID:SIGNAL:SECURITY. NM escapes colons in SSID
        # with backslash — split on un-escaped colon.
        parts: list[str] = []
        cur = ""
        i = 0
        while i < len(line):
            if line[i] == "\\" and i + 1 < len(line):
                cur += line[i + 1]; i += 2; continue
            if line[i] == ":":
                parts.append(cur); cur = ""; i += 1; continue
            cur += line[i]; i += 1
        parts.append(cur)
        if len(parts) < 3 or not parts[0]:
            continue
        ssid, signal_str, security = parts[0], parts[1], parts[2]
        try:
            signal = int(signal_str)
        except ValueError:
            continue
        if ssid not in best or signal > best[ssid]["signal"]:
            best[ssid] = {"ssid": ssid, "signal": signal, "security": security}
    return sorted(best.values(), key=lambda d: -d["signal"])


def _join_wifi(ssid: str, psk: str) -> tuple[bool, str]:
    """Attempt to connect. Returns (ok, error_message).

    We build the connection profile explicitly rather than relying on
    `nmcli dev wifi connect`, because that path reuses any pre-existing
    profile for the SSID — and a half-created/broken one (e.g. from a prior
    failed attempt) yields
    '802-11-wireless-security.key-mgmt: property is missing'. So: delete any
    stale profile for this SSID first, then create a fresh one with the
    security type pinned (wpa-psk), then bring it up."""
    # 1. Wipe any existing profile(s) with this con-name so we start clean.
    _run_nmcli("connection", "delete", ssid, timeout=15)

    if psk:
        # 2a. Secured network — create with key-mgmt explicitly set.
        rc, out, err = _run_nmcli(
            "connection", "add", "type", "wifi",
            "con-name", ssid, "ifname", "wlan0", "ssid", ssid,
            "wifi-sec.key-mgmt", "wpa-psk", "wifi-sec.psk", psk,
            timeout=30,
        )
        if rc != 0:
            return False, (err.strip() or out.strip() or f"add rc={rc}")
    else:
        # 2b. Open network — no security node at all.
        rc, out, err = _run_nmcli(
            "connection", "add", "type", "wifi",
            "con-name", ssid, "ifname", "wlan0", "ssid", ssid,
            timeout=30,
        )
        if rc != 0:
            return False, (err.strip() or out.strip() or f"add rc={rc}")

    # 3. Activate it.
    rc, out, err = _run_nmcli("connection", "up", ssid, timeout=60)
    if rc == 0:
        return True, ""
    # Clean up the failed profile so the next attempt starts fresh.
    _run_nmcli("connection", "delete", ssid, timeout=15)
    return False, (err.strip() or out.strip() or f"up rc={rc}")


def _persist_id_txt(device_id: str, password: str) -> tuple[bool, str]:
    """Atomically write id.txt with the user-chosen credentials.

    Tries ID_FILE (default /var/lib/elderly-care/id.txt) first; if that path
    isn't writable (e.g. the dir is root-owned and we run as the pipeline
    user), falls back to ~/.config/elderly-care/id.txt — which main.cpp's
    loadDeviceIdentity() also checks. setup.sh chowns /var/lib/elderly-care
    to the runtime user so the primary path normally works."""
    candidates = [ID_FILE,
                  Path.home() / ".config" / "elderly-care" / "id.txt"]
    last_err = ""
    for path in candidates:
        try:
            path.parent.mkdir(parents=True, exist_ok=True)
            tmp = path.with_suffix(".tmp")
            tmp.write_text(f"{device_id}\n{password}\n")
            tmp.chmod(0o600)
            tmp.rename(path)
            log.info("id.txt written to %s", path)
            return True, ""
        except Exception as e:
            last_err = f"{type(e).__name__}: {e}"
            continue
    return False, last_err


# ── GATT callbacks ─────────────────────────────────────────────────────────
# Bless calls read/write handlers synchronously from the dbus thread. For
# anything that takes >100ms (nmcli scan, wifi connect) we hand off to the
# asyncio loop via call_soon_threadsafe + run_in_executor so the BLE thread
# isn't blocked.

def _on_read(char: BlessGATTCharacteristic, **kwargs) -> bytearray:
    uuid = char.uuid.lower()
    if uuid == CHAR_WIFI_SCAN.lower():
        # Synchronous — short enough (nmcli takes ~5 s); BLE peer waits.
        nets = _scan_wifi()
        return bytearray(json.dumps(nets, ensure_ascii=False).encode("utf-8"))
    if uuid == CHAR_STATUS.lower():
        return bytearray(_status_value)
    if uuid == CHAR_DONE.lower():
        return bytearray(_done_value)
    return bytearray()


def _notify_status(value: bytes) -> None:
    global _status_value
    _status_value = value
    log.info("status → %s", value.decode("utf-8", errors="replace"))
    try:
        _server.get_characteristic(CHAR_STATUS).value = bytearray(value)
        _server.update_value(SVC_UUID, CHAR_STATUS)
    except Exception as e:
        log.warning("status notify failed: %s", e)


def _notify_done(value: bytes) -> None:
    global _done_value, _setup_completed
    _done_value = value
    _setup_completed = True
    log.info("done → %s", value.decode("utf-8", errors="replace"))
    try:
        _server.get_characteristic(CHAR_DONE).value = bytearray(value)
        _server.update_value(SVC_UUID, CHAR_DONE)
    except Exception as e:
        log.warning("done notify failed: %s", e)


async def _handle_wifi_join(payload: bytes) -> None:
    try:
        body = json.loads(payload.decode("utf-8"))
    except Exception:
        _notify_status(b"err: bad json")
        return
    ssid = body.get("ssid")
    psk  = body.get("psk", "")
    if not ssid:
        _notify_status(b"err: ssid required")
        return
    _notify_status(b"connecting")
    ok, err = await asyncio.get_running_loop().run_in_executor(
        None, _join_wifi, ssid, psk
    )
    if ok:
        _notify_status(b"wifi_ok")
    else:
        _notify_status(("wifi_fail: " + err).encode("utf-8")[:200])


async def _handle_commit(payload: bytes) -> None:
    try:
        body = json.loads(payload.decode("utf-8"))
    except Exception:
        _notify_done(b"err: bad json")
        return
    device_id = body.get("device_id")
    password  = body.get("password")
    if not device_id or not password:
        _notify_done(b"err: device_id and password required")
        return
    if len(password) < 32:
        _notify_done(b"err: password must be >= 32 chars")
        return
    ok, err = await asyncio.get_running_loop().run_in_executor(
        None, _persist_id_txt, device_id, password
    )
    _notify_done(b"ok" if ok else ("err: " + err).encode("utf-8")[:200])


def _on_write(char: BlessGATTCharacteristic, value: bytearray, **kwargs) -> None:
    global _client_connected
    _client_connected = True   # write ⇒ a client is attached
    uuid = char.uuid.lower()
    log.info("write → %s (%d bytes)", uuid[-4:], len(value))
    if uuid == CHAR_WIFI_JOIN.lower():
        asyncio.run_coroutine_threadsafe(_handle_wifi_join(bytes(value)), _loop)
    elif uuid == CHAR_COMMIT.lower():
        asyncio.run_coroutine_threadsafe(_handle_commit(bytes(value)), _loop)
    else:
        log.warning("unknown write to %s", char.uuid)


# ── Main loop ──────────────────────────────────────────────────────────────

async def main() -> int:
    global _server, _loop
    _loop = asyncio.get_running_loop()
    _server = BlessServer(name=DEVICE_NAME, loop=_loop)
    _server.read_request_func  = _on_read
    _server.write_request_func = _on_write

    await _server.add_new_service(SVC_UUID)

    char_specs = [
        (CHAR_WIFI_SCAN, GATTCharacteristicProperties.read,
         GATTAttributePermissions.readable, b"[]"),
        (CHAR_WIFI_JOIN, GATTCharacteristicProperties.write,
         GATTAttributePermissions.writeable, b""),
        (CHAR_STATUS,
         GATTCharacteristicProperties.read | GATTCharacteristicProperties.notify,
         GATTAttributePermissions.readable, b"idle"),
        (CHAR_COMMIT, GATTCharacteristicProperties.write,
         GATTAttributePermissions.writeable, b""),
        (CHAR_DONE,
         GATTCharacteristicProperties.read | GATTCharacteristicProperties.notify,
         GATTAttributePermissions.readable, b""),
    ]
    for uuid, props, perms, initial in char_specs:
        await _server.add_new_characteristic(
            SVC_UUID, uuid, props, bytearray(initial), perms)

    await _server.start()
    log.info("advertising as %s", DEVICE_NAME)
    log.info("service UUID: %s", SVC_UUID)

    # Window A — exit if no client connects within 5 minutes.
    waited = 0.0
    while not _client_connected and waited < WINDOW_A_SECONDS:
        await asyncio.sleep(1)
        waited += 1.0
    if not _client_connected:
        log.info("no client in %d s — exiting", WINDOW_A_SECONDS)
        await _server.stop()
        return 1

    # Window B — stay alive until commit completes (or signal).
    log.info("client attached; waiting for commit_setup")
    while not _setup_completed:
        await asyncio.sleep(1)

    # Brief tail so the final `done` notification flushes.
    await asyncio.sleep(2)
    log.info("commit complete — shutting down")
    await _server.stop()
    return 0


def _signal_handler(signum, frame):  # noqa: ARG001
    global _setup_completed
    log.info("signal %d — exiting", signum)
    _setup_completed = True


if __name__ == "__main__":
    signal.signal(signal.SIGTERM, _signal_handler)
    signal.signal(signal.SIGINT,  _signal_handler)
    sys.exit(asyncio.run(main()))

File: App/lib/services/ble_provisioning.dart
// Phase 7 — BLE provisioning client (companion side).
//
// Scans for the device's GATT service ("ElderlyCare-Setup"), connects,
// reads the Wi-Fi scan list, sends chosen credentials, then sends the
// final {device_id, password} commit. Mirrors the topic + payload shapes
// in device-side ble_provisioning.py.
//
// Usage pattern (called from setup_new_device_page.dart):
//   final svc = BleProvisioningService();
//   await svc.scanAndConnect();
//   final networks = await svc.scanWifi();
//   await svc.joinWifi(ssid, psk);
//   await svc.commitSetup(deviceId, password);
//   await svc.disconnect();
//
// Status updates are exposed via the `status` Stream so the UI can show
// "scanning", "connecting", "wifi_ok", etc. without polling.
import 'dart:async';
import 'dart:convert';

import 'package:flutter_blue_plus/flutter_blue_plus.dart';

/// One discovered provisioning device (for the picker). [name] is the
/// advertised "ElderlyCare-Setup-<serial>"; [rssi] lets us sort closest-first.
class DiscoveredDevice {
  final BluetoothDevice device;
  final String name;
  final int rssi;
  DiscoveredDevice({required this.device, required this.name, required this.rssi});
}

class BleProvisioningService {
  // Must match ble_provisioning.py's UUIDs byte-for-byte.
  static final Guid svcUuid       = Guid('11111111-2222-3333-4444-555555555555');
  static final Guid charWifiScan  = Guid('11111111-2222-3333-4444-555555555001');
  static final Guid charWifiJoin  = Guid('11111111-2222-3333-4444-555555555002');
  static final Guid charStatus    = Guid('11111111-2222-3333-4444-555555555003');
  static final Guid charCommit    = Guid('11111111-2222-3333-4444-555555555004');
  static final Guid charDone      = Guid('11111111-2222-3333-4444-555555555005');

  // Must match ble_provisioning.py's DEVICE_NAME prefix ("Noban-Setup-...").
  static const String devicePrefix = 'Noban';

  BluetoothDevice? _device;
  final Map<Guid, BluetoothCharacteristic> _chars = {};
  // GATT notification subscriptions — cancelled on re-bind + disconnect so a
  // reconnect doesn't stack duplicate listeners and a late notification can't
  // fire status.add() after the controller is closed ("add after close").
  StreamSubscription<List<int>>? _statusCharSub;
  StreamSubscription<List<int>>? _doneCharSub;

  /// Emits human-readable status messages — "scanning", "connecting", and
  /// every notification from the device's `status` characteristic
  /// ("connecting", "wifi_ok", "wifi_fail: …", etc.). UI can listen and
  /// render. Stream is closed on disconnect().
  final StreamController<String> status = StreamController.broadcast();

  /// Emits the device's `done` notification once `commitSetup` completes.
  /// Either "ok" or "err: ...".
  final StreamController<String> done = StreamController.broadcast();

  /// Scan for nearby devices advertising svcUuid, pick the first match,
  /// pair (Just Works), discover services, cache characteristic handles.
  /// Throws TimeoutException if no device found within `scanTimeout`.
  /// Scan for all in-range provisioning devices and return them sorted by
  /// signal strength (closest first). Each entry carries the advertised
  /// name ("ElderlyCare-Setup-<serial>") so the user can pick the right one
  /// when several are advertising. De-duplicates by remote id.
  Future<List<DiscoveredDevice>> scanForDevices({
    Duration scanTimeout = const Duration(seconds: 6),
  }) async {
    status.add('scanning');
    // De-dup by advertised NAME (now unique per device: ElderlyCare-Setup-
    // <serial>). The Pi advertises on two BT addresses (public + a rotating
    // random/RPA), so keying by remoteId would list one physical device
    // twice. Keying by name collapses them; keep the strongest-signal entry
    // (that's the address we'll actually connect to).
    final byKey = <String, DiscoveredDevice>{};
    final sub = FlutterBluePlus.scanResults.listen((results) {
      for (final r in results) {
        final adv = r.advertisementData.advName;
        // Only list adverts that carry the full serial form
        // "ElderlyCare-Setup-XXXX". This excludes (a) the adapter's own name
        // ("pi"), (b) UUID-only adverts with no/empty name, and (c) stale
        // pre-serial "ElderlyCare-Setup" adverts cached from earlier runs —
        // all of which previously showed up as a phantom second device.
        if (!adv.startsWith('$devicePrefix-')) continue;
        // Dedup by name (unique per device). The Pi advertises on two BT
        // addresses; keep the strongest-signal one — that's what we connect to.
        final existing = byKey[adv];
        if (existing == null || r.rssi > existing.rssi) {
          byKey[adv] = DiscoveredDevice(device: r.device, name: adv, rssi: r.rssi);
        }
      }
    });
    await FlutterBluePlus.startScan(withServices: [svcUuid], timeout: scanTimeout);
    await Future.delayed(scanTimeout);
    await sub.cancel();
    await FlutterBluePlus.stopScan();
    final list = byKey.values.toList()..sort((a, b) => b.rssi.compareTo(a.rssi));
    status.add(list.isEmpty ? 'no_devices' : 'found');
    return list;
  }

  /// Connect to a specific device chosen from [scanForDevices], discover the
  /// provisioning service + characteristics, and subscribe to notifications.
  Future<void> connectTo(BluetoothDevice device) async {
    _device = device;
    status.add('connecting');
    await _device!.connect(autoConnect: false, timeout: const Duration(seconds: 15));
    // NOTE: do NOT requestMtu() here. The writes (wifi_join ~42 B, commit
    // ~65 B) go through fine at the default MTU via ATT long writes, and
    // forcing an MTU change mid-session broke the wifi_scan characteristic
    // read (returned empty). The earlier "commit never arrived" was a UI
    // _busy bug, not an MTU problem.
    final services = await _device!.discoverServices();
    await _bindService(services);
  }

  /// Convenience: scan and connect to the strongest device (single-device
  /// homes). Throws TimeoutException if none found.
  Future<void> scanAndConnect({
    Duration scanTimeout = const Duration(seconds: 8),
  }) async {
    final found = await scanForDevices(scanTimeout: scanTimeout);
    if (found.isEmpty) {
      throw TimeoutException('no ElderlyCare-Setup device found');
    }
    await connectTo(found.first.device);
  }

  Future<void> _bindService(List<BluetoothService> services) async {
    final svc = services.firstWhere(
      (s) => s.uuid == svcUuid,
      orElse: () => throw StateError('provisioning service not found'),
    );
    for (final c in svc.characteristics) {
      _chars[c.uuid] = c;
    }

    // Subscribe to status + done notifications, forward into our streams.
    // Cancel any prior subscriptions first (a retried connect re-binds).
    await _statusCharSub?.cancel();
    await _doneCharSub?.cancel();
    final st = _chars[charStatus];
    if (st != null) {
      await st.setNotifyValue(true);
      _statusCharSub = st.lastValueStream.listen((bytes) {
        final s = utf8.decode(bytes, allowMalformed: true);
        if (s.isNotEmpty && !status.isClosed) status.add(s);
      });
    }
    final dn = _chars[charDone];
    if (dn != null) {
      await dn.setNotifyValue(true);
      _doneCharSub = dn.lastValueStream.listen((bytes) {
        final s = utf8.decode(bytes, allowMalformed: true);
        if (s.isNotEmpty && !done.isClosed) done.add(s);
      });
    }
    status.add('connected');
  }

  /// Read the Wi-Fi scan characteristic and parse the JSON list of
  /// {ssid, signal, security} maps. Sorted by signal strength descending.
  Future<List<Map<String, dynamic>>> scanWifi() async {
    final c = _chars[charWifiScan];
    if (c == null) throw StateError('not connected');
    final bytes = await c.read();
    try {
      final j = jsonDecode(utf8.decode(bytes));
      if (j is List) return j.cast<Map<String, dynamic>>();
    } catch (_) {}
    return const [];
  }

  /// Write the chosen Wi-Fi credentials. The device's status characteristic
  /// will emit "connecting" then either "wifi_ok" or "wifi_fail: ...".
  Future<void> joinWifi(String ssid, String psk) async {
    final c = _chars[charWifiJoin];
    if (c == null) throw StateError('not connected');
    final body = utf8.encode(jsonEncode({'ssid': ssid, 'psk': psk}));
    await c.write(body, withoutResponse: false);
  }

  /// Write the final device-id + 32+ char password. The device's `done`
  /// characteristic emits "ok" or "err: ...". Device will exit the sidecar
  /// shortly after.
  Future<void> commitSetup(String deviceId, String password) async {
    final c = _chars[charCommit];
    if (c == null) throw StateError('not connected');
    final body = utf8.encode(jsonEncode({
      'device_id': deviceId,
      'password':  password,
    }));
    await c.write(body, withoutResponse: false);
  }

  Future<void> disconnect() async {
    // Cancel notification subs BEFORE closing controllers so a late
    // notification can't call add() on a closed controller.
    await _statusCharSub?.cancel();
    await _doneCharSub?.cancel();
    _statusCharSub = null;
    _doneCharSub = null;
    try { await _device?.disconnect(); } catch (_) {}
    _device = null;
    _chars.clear();
    if (!status.isClosed) await status.close();
    if (!done.isClosed) await done.close();
  }
}
